File: Go/7/problem-1/problem-1-solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Evaluate an expression left-to-right, ignoring operator precedence
func evaluateExpression(expression string) int64 {
	var result, current int64
	var lastOperator byte = '+'

	for i := 0; i < len(expression); i++ {
		char := expression[i]
		if char >= '0' && char <= '9' {
			current = current*10 + int64(char-'0')
		} else if char == '+' || char == '*' {
			if lastOperator == '+' {
				result += current
			} else if lastOperator == '*' {
				result *= current
			}
			current = 0
			lastOperator = char
		}
	}

	// Final operation
	if lastOperator == '+' {
		result += current
	} else if lastOperator == '*' {
		result *= current
	}

	return result
}

// Generate all possible operator combinations and validate the result
func checkEquation(result int64, numbers []int) bool {
	combinations := 1 << (len(numbers) - 1) // Total combinations of '+' and '*'

	for i := 0; i < combinations; i++ {
		var expression strings.Builder
		for j, num := range numbers {
			expression.WriteString(strconv.Itoa(num))
			if j < len(numbers)-1 {
				if (i & (1 << j)) != 0 {
					expression.WriteByte('*')
				} else {
					expression.WriteByte('+')
				}
			}
		}

		if evaluateExpression(expression.String()) == result {
			fmt.Printf("Valid: %s = %d\n", expression.String(), result)
			return true
		}
	}

	return false
}

func main() {
	file, err := os.Open("inputdata.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var totalSum int64
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			continue
		}

		// Parse the result from the left-hand side
		result, err := strconv.ParseInt(strings.TrimSpace(parts[0]), 10, 64)
		if err != nil {
			continue
		}

		// Parse the numbers from the right-hand side
		numberStrings := strings.Fields(parts[1])
		numbers := make([]int, len(numberStrings))
		for i, numStr := range numberStrings {
			num, err := strconv.Atoi(numStr)
			if err != nil {
				continue
			}
			numbers[i] = num
		}

		// Check the equation
		if checkEquation(result, numbers) {
			totalSum += result
		}
	}

	fmt.Printf("Total sum of proven true results: %d\n", totalSum)
}

File: 2024/C/7/problem-1/problem-1-solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to evaluate the expression left-to-right (ignoring operator precedence)
long long evaluateExpression(char* expression) {
    long long result = 0;
    long long current = 0;
    char lastOperator = '+';

    for (int i = 0; expression[i]; i++) {
        if (expression[i] >= '0' && expression[i] <= '9') {
            current = current * 10 + (expression[i] - '0');
        } else if (expression[i] == '+' || expression[i] == '*') {
            if (lastOperator == '+') {
                result += current;
            } else if (lastOperator == '*') {
                result *= current;
            }
            current = 0;
            lastOperator = expression[i];
        }
    }

    // Final operation
    if (lastOperator == '+') {
        result += current;
    } else if (lastOperator == '*') {
        result *= current;
    }

    return result;
}

// Function to check all possible operator combinations
int checkEquation(long long result, int* numbers, int count) {
    int combinations = 1 << (count - 1); // Total combinations of '+' and '*'
    char expression[256];

    for (int i = 0; i < combinations; i++) {
        int pos = 0;
        for (int j = 0; j < count; j++) {
            pos += sprintf(expression + pos, "%d", numbers[j]);
            if (j < count - 1) {
                expression[pos++] = (i & (1 << j)) ? '*' : '+';
            }
        }
        expression[pos] = '\0';

        if (evaluateExpression(expression) == result) {
            printf("Valid: %s = %lld\n", expression, result);
            return 1; // Valid equation found
        }
    }

    return 0; // No valid equation found
}

int main() {
    FILE* file = fopen("inputdata.txt", "r");
    if (!file) {
        printf("Error opening file.\n");
        return 1;
    }

    long long totalSum = 0;
    char line[1024];

    while (fgets(line, sizeof(line), file)) {
        long long result;
        char* rightSide;
        int numbers[100];
        int count = 0;

        // Parse the line
        char* colon = strchr(line, ':');
        if (!colon) continue;

        *colon = '\0';
        result = atoll(line);
        rightSide = colon + 2; // Skip ": "

        // Parse numbers
        char* token = strtok(rightSide, " ");
        while (token) {
            numbers[count++] = atoi(token);
            token = strtok(NULL, " ");
        }

        // Check the equation
        if (checkEquation(result, numbers, count)) {
            totalSum += result; // Add the result from the left-hand side
        }
    }

    fclose(file);

    printf("Total sum of proven true results: %lld\n", totalSum);

    return 0;
}
